### PR TITLE
[SCEV] Add initial matchers for SCEV expressions. (NFC)

### PR DESCRIPTION
**Pull request by fhahn**
This patch adds initial matchers for SCEV expressions with an arbitrary
number of operands and specializes it for binary add expressions.

Also adds matchers for SCEVConstant and SCEVUnknown.

This patch only converts a few instances to use the new matchers to make
sure everything builds as expected for now.

The goal of the matchers is to hopefully make it slightly easier to write
code matching SCEV patterns.

Depends on https://github.com/llvm/llvm-project/pull/119389

### CODE REVIEW

**Comment by nikic** [comment]
Needs a rebase.

**Comment by fhahn** [comment]
Rebased after https://github.com/llvm/llvm-project/pull/119389

Also added some unary matchers to make use of the flexibility of `SCEVExpr_match`

File: llvm/include/llvm/Analysis/ScalarEvolutionPatternMatch.h [hunk 1]
```diff
@@ -52,6 +52,120 @@ inline cst_pred_ty<is_all_ones> m_scev_AllOnes() {
...
+    });
+  }
+};
+
+template <typename Op0_t, typename Op1_t, typename SCEVTy>
+using BinarySCEVExpr_match = SCEVExpr_match<std::tuple<Op0_t, Op1_t>, SCEVTy>;
+
+template <typename Op0_t, typename Op1_t, typename SCEVTy>
+inline BinarySCEVExpr_match<Op0_t, Op1_t, SCEVTy>
+m_scev_Binary(const Op0_t &Op0, const Op1_t &Op1) {
```
Comment (nikic) [inline llvm/include/llvm/Analysis/ScalarEvolutionPatternMatch.h:139]: How are you supposed to actually use this one? Don't you have to make SCEVTy the first template parameter to make this ergnomic?

File: llvm/include/llvm/Analysis/ScalarEvolutionPatternMatch.h [hunk 2]
```diff
@@ -52,6 +52,120 @@ inline cst_pred_ty<is_all_ones> m_scev_AllOnes() {
...
+inline BinarySCEVExpr_match<Op0_t, Op1_t, SCEVAddExpr>
+m_scev_Add(const Op0_t &Op0, const Op1_t &Op1) {
+  return BinarySCEVExpr_match<Op0_t, Op1_t, SCEVAddExpr>(Op0, Op1);
+}
+
+template <typename Op0_t, typename SCEVTy>
+using UnarySCEVExpr_match = SCEVExpr_match<std::tuple<Op0_t>, SCEVTy>;
+
+template <typename Op0_t, typename Op1_t, typename SCEVTy>
+inline UnarySCEVExpr_match<Op0_t, SCEVTy> m_scev_Unary(const Op0_t &Op0) {
```
Comment (nikic) [inline llvm/include/llvm/Analysis/ScalarEvolutionPatternMatch.h:153]: Same question here.

**Review by nikic — COMMENTED** [review]
This looks reasonable, but I can't claim to understand the template magic going on here. I do wonder if it wouldn't be better to just have separate unary and binary implementations and not try to be overly generic.

**Comment by fhahn** [comment]
> This looks reasonable, but I can't claim to understand the template magic going on here. I do wonder if it wouldn't be better to just have separate unary and binary implementations and not try to be overly generic.

Most of the complexity is coming from a missing utility to iterate over all members of a tuple. Happy to change it to a non-generic version. My original thought was that it might be convenient to match SCEV expressions with arbitrary number of arguments, but that would probably not be very common.

**Comment by fhahn** [comment]
Updated to have dedicated matchers for unary and binary exprs

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**